File: services/identity-service/src/test/java/br/com/deltaglobalbank/identity/features/signingKeys/rotateSigningKeyUseCase/SigningKeyRotatorTests.java
package br.com.deltaglobalbank.identity.features.signingKeys.rotateSigningKeyUseCase;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.security.KeyPair;

import br.com.deltaglobalbank.identity.domain.token.SigningKey;
import br.com.deltaglobalbank.identity.domain.token.SigningKeyRepository;
import br.com.deltaglobalbank.identity.domain.token.SigningKeyStatus;
import br.com.deltaglobalbank.identity.features.signingKeys.rotateSigningKeys.SigningKeyRotator;
import br.com.deltaglobalbank.identity.infrastructure.security.jwt.KeyGenerator;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class SigningKeyRotatorTests {

    private final SigningKeyRepository signingKeyRepository = mock(SigningKeyRepository.class);
    private final KeyGenerator keyGenerator = mock(KeyGenerator.class);

    private SigningKeyRotator rotator;

    @BeforeEach
    void setUp() {
        rotator = new SigningKeyRotator(signingKeyRepository, keyGenerator);

        KeyPair fakePair = mock(KeyPair.class);
        when(keyGenerator.generateRsaKeyPair()).thenReturn(fakePair);
        when(keyGenerator.encodePublicKey(any())).thenReturn("pub-new");
        when(keyGenerator.encodePrivateKey(any())).thenReturn("priv-new");

        when(signingKeyRepository.save(any())).thenAnswer(inv -> inv.getArgument(0));
    }

    @Test
    void mustRetireThePreviousKeyAndCreateANewActiveKey() {
        SigningKey previous = SigningKey.create("key-old", "RS256", "pub-old", "priv-old");
        when(signingKeyRepository.findFirstActive()).thenReturn(previous);

        SigningKeyRotator.RotationResult result = rotator.rotate();
        SigningKey newKey = result.newKey();
        SigningKey prev = result.previousKey();

        Assertions.assertAll(
            () -> assertEquals(SigningKeyStatus.RETIRED, prev.status()),
            () -> assertEquals(SigningKeyStatus.ACTIVE, newKey.status()),
            () -> assertNotEquals(prev.getKid(), newKey.getKid())
        );

        verify(signingKeyRepository, times(2)).save(any());
    }

    @Test
    void mustCreateANewActiveKeyWithoutRetiringWhenThereIsNoPreviousKey() {
        when(signingKeyRepository.findFirstActive()).thenReturn(null);

        SigningKeyRotator.RotationResult result = rotator.rotate();

        Assertions.assertAll(
            () -> assertNull(result.previousKey()),
            () -> assertEquals(SigningKeyStatus.ACTIVE, result.newKey().status())
        );

        verify(signingKeyRepository, times(1)).save(any());
    }
}

File: services/identity-service/src/test/kotlin/br/com/deltaglobalbank/identity/features/signingKeys/rotateSigningKeyUseCase/SigningKeyRotatorTests.kt
package br.com.deltaglobalbank.identity.features.signingKeys.rotateSigningKeyUseCase

import br.com.deltaglobalbank.identity.domain.token.SigningKey
import br.com.deltaglobalbank.identity.domain.token.SigningKeyRepository
import br.com.deltaglobalbank.identity.domain.token.SigningKeyStatus
import br.com.deltaglobalbank.identity.features.signingKeys.rotateSigningKeys.SigningKeyRotator
import br.com.deltaglobalbank.identity.infrastructure.security.jwt.KeyGenerator
import io.mockk.every
import io.mockk.verify
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.mockk
import org.junit.jupiter.api.Assertions.assertAll
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotEquals
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.extension.ExtendWith
import kotlin.test.Test

@ExtendWith(MockKExtension::class)
class SigningKeyRotatorTests {

    @MockK
    lateinit var signingKeyRepository: SigningKeyRepository

    @MockK
    lateinit var keyGenerator: KeyGenerator

    private lateinit var rotator: SigningKeyRotator

    @BeforeEach
    fun setUp() {
        rotator = SigningKeyRotator(signingKeyRepository, keyGenerator)

        val fakePair = mockk<java.security.KeyPair>(relaxed = true)
        every { keyGenerator.generateRsaKeyPair() } returns fakePair
        every { keyGenerator.encodePublicKey(any()) } returns "pub-new"
        every { keyGenerator.encodePrivateKey(any()) } returns "priv-new"

        every { signingKeyRepository.save(any()) } answers { firstArg() }
    }

    @Test
    fun `must retire the previous key and create a new active key`() {
        val previous = SigningKey.create("key-old", "RS256", "pub-old", "priv-old")
        every { signingKeyRepository.findFirstActive() } returns previous

        val (newKey, prev) = rotator.rotate()

        assertAll(
            { assertEquals(SigningKeyStatus.RETIRED, prev!!.status()) },
            { assertEquals(SigningKeyStatus.ACTIVE, newKey.status()) },
            { assertNotEquals(prev!!.kid, newKey.kid) }
        )

        verify(exactly = 2) { signingKeyRepository.save(any()) }
    }

    @Test
    fun `must create a new active key without retiring when there is no previous key`() {
        every { signingKeyRepository.findFirstActive() } returns null

        val (newKey, prev) = rotator.rotate()

        assertAll(
            { assertNull(prev) },
            { assertEquals(SigningKeyStatus.ACTIVE, newKey.status()) }
        )

        verify(exactly = 1) { signingKeyRepository.save(any()) }
    }
}